199 UI test add a vehicle

am2089 opened the pull request:
# What it Does
* Closes #199 
* UI Test for add a vehicle in settings

# How I Tested
Made sure all tests passed before committing

am2089 committed:
List ordered by amount of contributions and profile name

am2089 committed:
Changes made topbxproj, and moved code from the contributorslistview to settingsviewmodel

am2089 committed:
changes requested

am2089 committed:
Changes

am2089 committed:
Committed changes to Basic-Car-Maintenance.xcodeproj/project.pbxproj

am2089 committed:
UITest for adding a vehicle in settings

mikaelacaron commented on Basic-Car-Maintenance.xcodeproj/project.pbxproj:
```diff
@@ -219,8 +221,8 @@
 			children = (
-				FFBE79BA2AD0A48C0005524E /* Configurations */,
 				FF098EFA2AB3424E003EC0FE /* Basic-Car-Maintenance.xcconfig */,
+				FFBE79BA2AD0A48C0005524E /* Configurations */,
```
This shouldn't be changed

mikaelacaron commented on Basic-Car-Maintenance.xcodeproj/project.pbxproj:
```diff
@@ -662,6 +666,7 @@
 				ALWAYS_SEARCH_USER_PATHS = NO;
 				ASSETCATALOG_COMPILER_GENERATE_SWIFT_ASSET_SYMBOL_EXTENSIONS = YES;
+				CLANG_ANALYZER_LOCALIZABILITY_NONLOCALIZED = YES;
 				CLANG_ANALYZER_NONNULL = YES;
```
Do you know what the purpose of this setting is? and why it needs to be set to yes?

mikaelacaron commented on Basic-Car-Maintenance.xcodeproj/project.pbxproj:
```diff
@@ -691,6 +696,7 @@
 				CLANG_WARN_UNREACHABLE_CODE = YES;
 				CLANG_WARN__DUPLICATE_METHOD_MATCH = YES;
 				COPY_PHASE_STRIP = NO;
+				DEAD_CODE_STRIPPING = YES;
```
Do you know what the purpose of this setting is? and why it needs to be set to yes?

mikaelacaron commented on Basic-Car-Maintenance.xcodeproj/project.pbxproj:
```diff
@@ -774,12 +782,14 @@
 		};
 		FF5D13C92A86C2D800BC9BD6 /* Debug */ = {
 			isa = XCBuildConfiguration;
+			baseConfigurationReference = FF098EFA2AB3424E003EC0FE /* Basic-Car-Maintenance.xcconfig */;
```
This should not be changed. Re-read the README again to properly set up this project

mikaelacaron commented on Basic-Car-Maintenance.xcodeproj/project.pbxproj:
```diff
@@ -840,6 +851,8 @@
 				"LD_RUNPATH_SEARCH_PATHS[sdk=macosx*]" = "@executable_path/../Frameworks";
 				MACOSX_DEPLOYMENT_TARGET = 13.4;
 				OTHER_LDFLAGS = "-ObjC";
+				"PRODUCT_BUNDLE_IDENTIFIER[sdk=iphoneos*]" = "com.andrewmuniz.Basic-Car-Maintenance";
+				"PRODUCT_BUNDLE_IDENTIFIER[sdk=macosx*]" = "com.andrewmuniz.Basic-Car-Maintenance";
```
This should not be changed

mikaelacaron commented on Basic-Car-Maintenance/Shared/Settings/ViewModels/SettingsViewModel.swift:
```diff
@@ -16,6 +16,23 @@ final class SettingsViewModel {
     
     var vehicles = [Vehicle]()
     
+    var sortedContributors: [Contributor] {
+        guard let contributors = contributors, !contributors.isEmpty else {
+            return []
+        }
+        
+        return contributors.sorted { (contributor1, contributor2) in
+            switch (contributor1.contributions, contributor2.contributions) {
+            case let (contributor1, contributor2) where contributor1 > contributor2:
+                return true
+            case let (contributor1, contributor2) where contributor1 < contributor2:
+                return false
+            default:
+                return contributor1.login < contributor2.login
+            }
+        }
+    }
+    
```
This shouldn't be changed at all, you need to update your fork before making a new branch for new issues

mikaelacaron commented on Basic-Car-Maintenance/Shared/Settings/Views/ContributorsListView.swift:
```diff
@@ -8,13 +8,12 @@
         List {
-            if let contributors = viewModel.contributors, !contributors.isEmpty {
-                ForEach(contributors) { contributor in
+            if !viewModel.sortedContributors.isEmpty {
+                ForEach(viewModel.sortedContributors) { contributor in
```
This shouldn't be changed

mikaelacaron commented on Basic-Car-Maintenance-UITests/AddVehicleViewUITest.swift:
```diff
@@ -0,0 +1,41 @@
+        app.buttons["Add Vehicle"].tap()
+        
+        // Fill in valid vehicle information
+        let nameTextField = app.textFields["Vehicle Name"]
```
This test failed for me:
> Failed to tap "Vehicle Name" TextField: No matches found for Descendants matching type TextField from input

mikaelacaron reviewed (changes requested):
Good work! Note I did specifically say in the issue to
> Just add it directly to the `BasicCarMaintenanceUITests` file for now, no need to make a new one

Which this has caused a merge conflict that needs to be resolved, also I don't think you updated your fork, or what branch you worked on. There's changes in the `SettingsViewModel` that shouldn't be there.

Also note several files changed that I wouldn't expect, in the `xcschemes/WidgetExtension.xcscheme` and
`/xcschemes/Basic-Car-Maintenance-WidgetExtension.xcscheme`

Address these issues and click the re-review button when you're ready, comment on the thread if you have any questions

<img width="345" alt="Re-review button" src="https://github.com/mikaelacaron/Basic-Car-Maintenance/assets/22946902/c4d7061e-97b5-46f8-b2b0-b73c14b078f3">

am2089 committed:
Updated Basic-Car-Maintenance-UITests

am2089 commented:
Hey, so i made sure my my fork was up to date now, and moved the UITest into the file you wanted. I think the pbx file was old before you made the changes and it was just a commit that was still there that got pushed with the UITest I made. I could be doing something wrong lol but I double checked that I am up to date now and just pushed the changed where I moved the code.

mikaelacaron reviewed (changes requested):
@am2089 there's still several changes that are there but should be and from looking at GitHub it's still saying your branch is **not** up to date with the latest in `dev`

am2089 commented:
Im just curious, should I just re-fork the project so it can be absolutely up to date? and just add the test I wrote, and create a new pull request so everything will be on a clean slate?

mikaelacaron commented:
> Im just curious, should I just re-fork the project so it can be absolutely up to date? and just add the test I wrote, and create a new pull request so everything will be on a clean slate?

No. You need to update the `dev` branch of you fork by clicking the sync button in GitHub, and then merge those changes into you branch. I did a livestream about this the other day. Start around `1:20:44` https://www.youtube.com/live/-rYYKXrCyro?si=J2IQNjJyqa6wiwrQ

am2089 commented:
So I did that before I even made the fork and the branch. 
![Screenshot 2023-10-22 at 9 12 12 PM](https://github.com/mikaelacaron/Basic-Car-Maintenance/assets/63330690/8b89b71b-20ce-457e-bd52-7c74b904e916)

am2089 commented:
Actually i see  whats wrong my branch isn't updated and is 7 commits ahead, 54 commits behind. So should I delete the seven commits, update the branch and re commit the changes?

mikaelacaron commented:
No don't need to delete any commits. You only need to update your branch, (merging the changes from `dev` into your branch `199-UITest-AddAVehicle`

am2089 committed:
Resolved merge conflicts in project.pbxproj

am2089 commented:
So I fixed the pbx file in vscode, saved it, and committed it. Is there anything else I need to do? Or is it what I did not showing? also should i start a new pr, or did you close it because it was accepted?

mikaelacaron commented:
> So I fixed the pbx file in vscode, saved it, and committed it. Is there anything else I need to do? Or is it what I did not showing? also should i start a new pr, or did you close it because it was accepted?

You fix the code on your branch, and then push the changes to GitHub. There was no need to close the PR.
I didn't close the PR, you did.

But you can either re-open this one, or make a new one, it doesn't matter as long as it's been updated as shown in my livestream that I linked earlier

mikaelacaron commented:
@am2089 There's still a bunch of changes that I noted that haven't been fixed.
I think the best thing to do would be to copy the contents of the `BasicCarMaintenanceUITests.swift` file for what you've changed, and start a new branch on your fork from `dev` (because it will be updated)

Apply your changes, and then make a new PR

mikaelacaron reviewed (changes requested):
And the only file that should be changed would be the `BasicCarMaintenanceUITests.swift`, _unless_ you can explain why any Xcode settings need to be changed

am2089 committed:
Merge branch 'dev' into 199-UITest-AddAVehicle

am2089 commented:
So I'm sort of confused about what to do. I understand you pointed out all those commits, but I only want to commit the commits I made on Oct 22. How do I get rid of the rest? Sorry wrapping my head around git and Git Hub when a file gets messed up is a task. Also on a side note will you be continuing this project passed October?

mikaelacaron commented:
> How do I get rid of the rest? Sorry wrapping my head around git and Git Hub when a file gets messed up is a task. Also on a side note will you be continuing this project passed October?


One thing to do is just make a new branch and re commit everything, that's the simplest!

otherwise, you need to use a text editor and manually remove the stuff that shouldn't be there. Actually "removing commits" is possible, but significantly more difficult than you would expect.

Also yes! I'll be continuing this project after October! I need to add a paywall and put features behind the paywall. I'm hoping to ship within the first few weeks of November! This project will still be open for people to contribute, but I imagine it'll significantly slow down after October is over. You're welcome to still contribute to it!


Easiest way to fix this:
* Make sure your fork is up to date
* Checkout a new branch from `dev`
* Move the one change you had (everything in the `BasicCarMaintenanceUITests.swift` file)
   * Literally copy and paste what your change is into the new branch and run it
* Commit that
* Change the PR at the top to that new branch (or create a new PR)
